Implement new mod updater framework.

pchote opened the pull request:
This PR implements the plumbing promised in #14600 and ports over the rules from `release-20180218` to current bleed.  I would like to think that we can extend this back to `release-20171014` in a future PR to provide continuity for modders on the Mod SDK (I have a separate commit that makes a start on this).  Depends on #14948.

The key points to know about the new system are:
* All changes to the mod rules are expected to define an update rule, even if it only lists the changes that a modder needs to make manually.
* The default commands for the updater list information about the changes; an explicit `--apply` command must be given before any changes are made.  This makes the updater our primary documentation for compatibility changes.
* Each update "rule" is considered independently, and can be applied individually (e.g. when working on bleed), or as part of an "update path" from a labelled tag to the current HEAD.
* If a rule throws an exception then its changes will be reverted and any other updates in the path will be applied.

I expect there will be some bugs to shake out before we merge this, so testing by third party mods would be appreciated.  I am deliberately not explaining how to use this (except to note that it has been renamed to `--update-mod` / `--update-map`), because one of the main goals of this new system is to be self-explanatory and relatively intuitive.

reaperrr commented on OpenRA.Mods.Common/UpdateRules/Rules/DefineSoundDefaults.cs:
```diff
@@ -0,0 +1,114 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
```
2018 (applies to all new files).

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/DefineSoundDefaults.cs:
```diff
@@ -0,0 +1,114 @@
+				+ "You may wish to explicitly define `" + replacement + "` at the following locations\n"
+				+ "if the sound has not already been inherited from a parent definition.\n";
+
+			foreach (var n in nodes)
```
I should split out a helper for aggregating lines into sub-lists.  This is going to be done a lot.

abcdefg30 commented:
#14948 was merged.

abcdefg30 commented on OpenRA.Mods.Common/UtilityCommands/UpdateMapCommand.cs:
```diff
@@ -0,0 +1,122 @@
+			}
+
+			if (args.Contains("--apply"))
+			{
```
Braces around a single-line statement.

abcdefg30 reviewed (dismissed):
Looks good to me otherwise from a first look. 👍 
The manual changes section can become a bit spammy, but I guess that is to be expected.

pchote commented:
Updated and rebased.

Yes, this is designed to be spammy: if the rules don't explain everything that a modder needs to do to restore compatibility then the rules aren't written well enough.

pchote committed:
Implement new mod/map updater framework.

pchote commented:
Updated again.  I removed the widget updating plumbing because I want to change that, and it doesn't have any uses in the current PR.  We can add that back when porting over the older rules.

reaperrr commented:
Tested locally with @KOYK's TDA mod, didn't run into any issues, all commands and source types seem to work fine 👍

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/c608bb3aa2db0aba8908714bff2be57da0756ae7...51e7e1e0c827786b10540ff06c574b37128d1439
